client/app/visualizations/counter/Renderer.jsx

```diff
@@ -0,0 +1,74 @@
+import { isFinite } from 'lodash';
+import React, { useState, useEffect } from 'react';
+import cx from 'classnames';
+import resizeObserver from '@/services/resizeObserver';
+import { RendererPropTypes } from '@/visualizations';
+
+import { getCounterData } from './utils';
+
+import './render.less';
+
+function getCounterStyles(scale) {
+  return {
+    OTransform: `scale(${scale})`,
+    MsTransform: `scale(${scale})`,
+    MozTransform: `scale(${scale})`,
+    WebkitTransform: `scale(${scale})`,
+    transform: `scale(${scale})`,
+  };
+}
+
+function getCounterScale(container) {
+  const inner = container.firstChild;
+  const scale = Math.min(container.offsetWidth / inner.offsetWidth, container.offsetHeight / inner.offsetHeight);
+  return Number(isFinite(scale) ? scale : 1).toFixed(2); // keep only two decimal places
+}
+
+export default function Renderer({ data, options, visualizationName }) {
+  const [scale, setScale] = useState('1.00');
+  const [container, setContainer] = useState(null);
+
+  useEffect(() => {
+    if (container) {
+      const unwatch = resizeObserver(container, () => {
+        setScale(getCounterScale(container));
+      });
+      return unwatch;
+    }
+  }, [container, setScale]);
+
+  useEffect(() => {
+    if (container) {
+      // update scaling when options or data change (new formatting, values, etc.
+      // may change inner container dimensions which will not be tracked by `resizeObserver`);
+      setScale(getCounterScale(container));
+    }
+  }, [data, options, container, setScale]);
+
+  const {
+    showTrend, trendPositive,
+    counterValue, counterValueTooltip,
+    targetValue, targetValueTooltip,
+    counterLabel,
+  } = getCounterData(data.rows, options, visualizationName);
+  return (
+    <div
+      className={cx(
+        'counter-visualization-container',
+        { 'trend-positive': showTrend && trendPositive, 'trend-negative': showTrend && !trendPositive },
+      )}
+    >
+      <div className="counter-visualization-content" ref={setContainer}>
```

Comment:
Why not use the dedicated `useRef`?

Reply:
Previously I had some issues passing a regular ref as a dependency to `useEffect` (it just ignored ref changes). This variant works perfectly

Reply:
[This](https://reactjs.org/docs/hooks-faq.html#how-can-i-measure-a-dom-node) seems relevant to cases where the DOM changes and you need to track it in the ref.